keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+          { "fieldPath": "status.nominatedNodeName"},
+        ]
+      },
+```
+
+#### Declarative field selector definitions on built-in types
+
+To make it easier and safer to add field selector definitions for built-in types
+in the future. We will support defining field selectors on API types using
+go struct tags.
+
+```go
+type PodSpec struct {
+  // NodeName ...
+  //
+  // +selectableField
+  NodeName string `json:...`
+
+  // Host networking ...
+  //
+  // +k8s:schema:selectableField:fieldPath="spec.hostNetwork",
+  // +k8s:schema:selectableField:valueIfAbsent="false"
+  // +k8s:schema:selectableField:description="Set to the value of the spec.securityContext.hostNetwork field"
+  // +optional
+  HostNetwork bool `json:...`
+}
+```
+
+String, bool and integer typed fields, and typerefs to these types will be
+supported and automatically cast to string.
+
+`valueIfAbsent` will be required unless all field path parts for this field are
+required or have a default.
+
+`fieldSelector` should only be set for legacy cases where the field selector
+name does not match the field path. When set, a description will be required.
+
+`description` is provided for the legacy cases that use `fieldSelector` and
+require further clarification. For all other cases, it should not be needed.
+
+`Pod.status.podIP[0]` is a special legacy case. We will handle it in code. We
+may need to add special go tag part to allow us to include declarative
+information in the go struct for discovery purposes while handling the
+implementation imperatively.
+
+### User Stories
+
+- CustomResourceDefinition author wishes to provide filtered access by a field
+  - Author adds the field to an allow list in the CustomResourceDefinition
+  - If field is not always present, CustomResourceDefinition author declares how absent fields are handled
+- User uses `--field-selector` in kubectl to filter a list response.
+  - User first needs to know what field selectors are available and what values the selectable fields might have.
+    - Places user might look: discovery, CRD resource.
+  - User filter by:
+    - a single field
+    - multiple fields
+    - both fields and labels
+- Controller watches custom resources while filtering with field selectors.
+  - Controller provides a FieldSelector string in ListOptions 
+
+### Notes/Constraints/Caveats (Optional)
+
+Discussions about limitations of field selectors:
+
+- https://github.com/kubernetes/kubernetes/issues/32946
+- https://github.com/kubernetes/kubernetes/issues/72196
+- https://github.com/kubernetes/kubernetes/issues/107053
+
+Problems with escaping:
+
+- https://github.com/kubernetes/kubernetes/pull/28112
+
+### Future work
+
+#### Index selected fields
+
+Label and field selectors may be accelerated in the future. Possible approaches:
+
+- etcd adds filter/index/symlink support in the future and Kubernetes leverages it
+- the watch cache is updated to index labels and field selectors
```

Comment:
watch cache already does this, but it's never been extended.  I remember a PR from @wojtek-t

Reply:
We have a single predefined index for pods now (on spec.nodeName). The code is generic so we can definitely add more indexes, but we don't have any logic that automatically adds any indexes.
I would treat it as a potential future work until we really see a good data proving that it's needed.